server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java

```diff
@@ -1987,6 +2103,61 @@ public Snapshot copySnapshot(CopySnapshotCmd cmd) throws StorageUnavailableExcep
         }
     }
 
+    private boolean canCopyOnPrimary(List<Long> poolIds, Snapshot snapshot) {
+        List<Long> poolsToBeRemoved = new ArrayList<>();
+        for (Long poolId : poolIds) {
+            PrimaryDataStore dataStore = (PrimaryDataStore) dataStoreMgr.getDataStore(poolId, DataStoreRole.Primary);
+            if (dataStore == null) {
+                poolsToBeRemoved.add(poolId);
+                continue;
+            }
+            SnapshotInfo snapshotInfo = snapshotFactory.getSnapshot(snapshot.getId(), poolId, DataStoreRole.Primary);
+            if (snapshotInfo != null) {
+                logger.debug(String.format("Snapshot [%s] already exist on pool [%s]", snapshot.getUuid(), dataStore.getName()));
+                continue;
+            }
+
+            VolumeVO volume = _volsDao.findById(snapshot.getVolumeId());
+            if (volume == null) {
+                poolsToBeRemoved.add(poolId);
+                continue;
+            }
+            if (!dataStore.getDriver().getCapabilities()
+                    .containsKey(DataStoreCapabilities.CAN_COPY_SNAPSHOT_BETWEEN_ZONES_AND_SAME_POOL_TYPE.toString())
+                    && dataStore.getPoolType() != volume.getPoolType()) {
+                poolsToBeRemoved.add(poolId);
+                logger.debug(String.format("The %s  does not support copy to %s between zones", dataStore.getPoolType(), volume.getPoolType()));
+            }
+        }
+        poolIds.removeAll(poolsToBeRemoved);
+        if (CollectionUtils.isEmpty(poolIds)) {
+            return false;
+        }
+        return true;
+    }
```

Comment:
this can be further distilled into smaller methods. Surely some of those can than be reused for the actual copy method.